crates/ra_cli/src/analysis_stats.rs

```diff
@@ -52,6 +52,9 @@ pub fn run(verbose: bool, memory_usage: bool, path: &Path, only: Option<&str>) -
     println!("Total modules found: {}", visited_modules.len());
     println!("Total declarations: {}", num_decls);
     println!("Total functions: {}", funcs.len());
+    println!("Parsing: {:?}, {}", analysis_time.elapsed(), ra_prof::memory_usage());
```

Comment:
```suggestion
    println!("Item Collection: {:?}, {}", analysis_time.elapsed(), ra_prof::memory_usage());
```

The code about works with `Modules`, and to get modules we run top-level name resolution (crate_def_map_query).

Reply:
That is, this indeed counts all the parsing, but it actually counts more than that.

Reply:
> That is, this indeed counts all the parsing, but it actually counts more than that.

That means that in the type inference part there should be no parsing involved, right?

Reply:
Yeah, I would think so.

Reply:
That's strange. I recorded a profile from just after the new message (adding an `io::stdin().read_line()`, and it looks like this:

```
5.58%  <rowan::green::GreenNode as core::hash::Hash>::hash
4.20%  _int_malloc
4.08%  _int_free
3.65%  ra_syntax::syntax_node::SyntaxTreeBuilder::finish_node
3.42%  ra_parser::parser::Parser::nth
3.39%  malloc
3.04%  alloc::sync::Arc<T>::drop_slow
2.81%  __memmove_avx_unaligned_erms
2.71%  ra_parser::event::process
2.58%  unlink_chunk.isra.0
2.24%  malloc_consolidate
2.24%  <smol_str::SmolStr as core::hash::Hash>::hash
1.82%  ra_syntax::parsing::text_tree_sink::TextTreeSink::do_token
1.80%  <ra_syntax::parsing::text_token_source::TextTokenSource as ra_parser::TokenSource>::lookahead_nth
1.69%  ra_rustc_lexer::<impl ra_rustc_lexer::cursor::Cursor>::advance_token
1.48%  <alloc::vec::Vec<T> as alloc::vec::SpecExtend<T,I>>::from_iter
1.34%  <core::iter::adapters::Map<I,F> as core::iter::traits::iterator::Iterator>::next
1.18%  cfree@GLIBC_2.2.5
1.18%  indexmap::map::IndexMap<K,V,S>::insert
1.14%  <rowan::cursor::SyntaxNode as core::ops::drop::Drop>::drop
1.08%  <core::iter::adapters::Map<I,F> as core::iter::traits::iterator::Iterator>::try_fold
1.03%  <salsa::derived::DerivedStorage<DB,Q,MP> as salsa::plumbing::QueryStorageOps<DB,Q>>::try_fetch
1.01%  rowan::cursor::FreeList::try_push
1.00%  smol_str::SmolStr::new
0.93%  alloc::vec::Vec<T>::extend_from_slice
0.87%  ra_parser::parser::Parser::do_bump
0.82%  ra_parser::parser::Marker::complete
0.76%  <alloc::rc::Rc<T> as core::ops::drop::Drop>::drop
0.65%  alloc::sync::Arc<T>::drop_slow
0.63%  <ra_syntax::parsing::text_tree_sink::TextTreeSink as ra_parser::TreeSink>::start_node
0.62%  <ra_syntax::algo::visit::VisCtx<V,N,F> as ra_syntax::algo::visit::VisitorCtx>::accept
0.62%  ra_syntax::parsing::lexer::tokenize
0.61%  salsa::runtime::local_state::LocalState<DB>::report_query_read
0.60%  <salsa::derived::DerivedStorage<DB,Q,MP> as salsa::plumbing::QueryStorageOps<DB,Q>>::try_fetch
0.57%  realloc
0.51%  <rowan::cursor::SyntaxNodeChildren as core::iter::traits::iterator::Iterator>::next
0.51%  <alloc::vec::Vec<T> as alloc::vec::SpecExtend<T,I>>::from_iter
0.50%  <chalk_ir::Parameter as chalk_ir::fold::Fold>::fold_with
0.50%  drop_bomb::RealBomb::new
```

So the parser still seems to dominate.

Reply:
Ah, we have an LRU cache for parse trees. Could you try disableing it with `env RA_LRU_SIZE=0`?

Reply:
Same:

```
5.58%  <rowan::green::GreenNode as core::hash::Hash>::hash
4.42%  _int_malloc
4.23%  _int_free
3.76%  ra_syntax::syntax_node::SyntaxTreeBuilder::finish_node
3.56%  ra_parser::parser::Parser::nth
3.50%  malloc
2.91%  alloc::sync::Arc<T>::drop_slow
2.85%  __memmove_avx_unaligned_erms
2.71%  ra_parser::event::process
2.47%  unlink_chunk.isra.0
2.23%  <smol_str::SmolStr as core::hash::Hash>::hash
2.19%  malloc_consolidate
1.86%  ra_rustc_lexer::<impl ra_rustc_lexer::cursor::Cursor>::advance_token
1.78%  ra_syntax::parsing::text_tree_sink::TextTreeSink::do_token
1.70%  <ra_syntax::parsing::text_token_source::TextTokenSource as ra_parser::TokenSource>::lookahead_nth
1.57%  <alloc::vec::Vec<T> as alloc::vec::SpecExtend<T,I>>::from_iter
1.38%  <core::iter::adapters::Map<I,F> as core::iter::traits::iterator::Iterator>::next
1.25%  cfree@GLIBC_2.2.5
1.22%  <rowan::cursor::SyntaxNode as core::ops::drop::Drop>::drop
1.20%  indexmap::map::IndexMap<K,V,S>::insert
1.07%  <core::iter::adapters::Map<I,F> as core::iter::traits::iterator::Iterator>::try_fold
0.99%  rowan::cursor::FreeList::try_push
0.98%  alloc::vec::Vec<T>::extend_from_slice
0.98%  smol_str::SmolStr::new
0.93%  <alloc::rc::Rc<T> as core::ops::drop::Drop>::drop
0.83%  <salsa::derived::DerivedStorage<DB,Q,MP> as salsa::plumbing::QueryStorageOps<DB,Q>>::try_fetch
0.83%  ra_parser::parser::Marker::complete
0.81%  ra_parser::parser::Parser::do_bump
0.66%  <ra_syntax::parsing::text_tree_sink::TextTreeSink as ra_parser::TreeSink>::start_node
0.64%  alloc::sync::Arc<T>::drop_slow
0.63%  <ra_syntax::algo::visit::VisCtx<V,N,F> as ra_syntax::algo::visit::VisitorCtx>::accept
0.58%  salsa::runtime::local_state::LocalState<DB>::report_query_read
0.57%  <salsa::derived::DerivedStorage<DB,Q,MP> as salsa::plumbing::QueryStorageOps<DB,Q>>::try_fetch
0.57%  ra_syntax::parsing::lexer::tokenize
0.54%  <rowan::cursor::SyntaxNodeChildren as core::iter::traits::iterator::Iterator>::next
0.52%  core::iter::adapters::ResultShunt<I,E>::process
0.52%  0xffffffffbd800b07
0.51%  realloc
0.51%  ra_rustc_lexer::<impl ra_rustc_lexer::cursor::Cursor>::ident
0.51%  <alloc::vec::Vec<T> as alloc::vec::SpecExtend<T,I>>::from_iter
```

Reply:
`RA_LRU_CAP=0` :smile: 

```
8.00%  _int_free
6.58%  unlink_chunk.isra.0
6.33%  _int_malloc
5.41%  malloc_consolidate
4.67%  alloc::sync::Arc<T>::drop_slow
4.20%  __memmove_avx_unaligned_erms
4.05%  malloc
2.32%  indexmap::map::IndexMap<K,V,S>::insert
2.18%  cfree@GLIBC_2.2.5
1.80%  <salsa::derived::DerivedStorage<DB,Q,MP> as salsa::plumbing::QueryStorageOps<DB,Q>>::try_fetch
1.80%  alloc::vec::Vec<T>::extend_from_slice
1.58%  <alloc::vec::Vec<T> as core::ops::drop::Drop>::drop
1.36%  core::iter::adapters::ResultShunt<I,E>::process
1.21%  <salsa::derived::DerivedStorage<DB,Q,MP> as salsa::plumbing::QueryStorageOps<DB,Q>>::try_fetch
1.21%  salsa::runtime::local_state::LocalState<DB>::report_query_read
1.12%  <chalk_ir::Parameter as chalk_ir::fold::Fold>::fold_with
1.07%  <chalk_ir::Ty as core::clone::Clone>::clone
0.94%  0xffffffffbd800b07
0.89%  <alloc::vec::Vec<T> as alloc::vec::SpecExtend<T,I>>::from_iter
0.88%  realloc
0.84%  <salsa::interned::InternedStorage<DB,Q> as salsa::plumbing::QueryStorageOps<DB,Q>>::try_fetch
0.72%  <ra_ide_api::db::__SalsaDatabaseKey as core::clone::Clone>::clone
0.69%  <alloc::vec::Vec<T> as core::ops::drop::Drop>::drop
0.69%  core::ptr::real_drop_in_place
0.63%  std::collections::hash::map::HashMap<K,V,S>::get
0.60%  <chalk_rust_ir::AssociatedTyValue as chalk_solve::clauses::program_clauses::ToProgramClauses>::to_program_clauses
0.59%  chalk_ir::fold::super_fold_ty
0.58%  <rowan::cursor::SyntaxNodeChildren as core::iter::traits::iterator::Iterator>::next
0.55%  chalk_ir::Binders<T>::map_ref
0.51%  <ra_ide_api::db::RootDatabase as salsa::plumbing::HasQueryGroup<ra_hir::db::HirDatabaseStorage>>::database_key
0.50%  <ra_ide_api::db::RootDatabase as salsa::plumbing::HasQueryGroup<ra_hir::db::DefDatabaseStorage>>::database_key
```